<a href="https://colab.research.google.com/github/ginttone/test_visuallization/blob/master/2_autompg_linearregression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 머신러닝

- 정보(데이터)단계<br>
dropna:info(), describe()<br>
fillna, replace:describe(), value_counts()<br>
 - 시각화 :통계 선택<br>
standard scaler
혹은 one hot encoding<br> 뭐로 할지 정하기

- 교육단계<br>
stadard scaler, get_dummies(one hot encoding)<br>
model learning<br>
check score<br>

- 서비스단계 <br>
pickle: dump,load<br>
recive data <br>
apply prediction<br>

## 데이터 로딩



In [1]:
import pandas as pd
df= pd.read_csv('./auto-mpg.csv', header=None)
df.columns=['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name']
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    float64
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


In [2]:
df[['horsepower','name']].describe(include='all')

horsepower        name
count         398         398
unique         94         305
top         150.0  ford pinto
freq           22           6

## replace

In [3]:
df['horsepower'].value_counts()

150.0    22
90.00    20
88.00    19
110.0    18
100.0    17
         ..
200.0     1
77.00     1
103.0     1
133.0     1
152.0     1
Name: horsepower, Length: 94, dtype: int64

In [4]:
df['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '?', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [5]:
df_horsepower=df['horsepower'].replace(to_replace='?',value=None,inplace=False)
df_horsepower.unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [6]:
df_horsepower=df_horsepower.astype('float')

In [7]:
df_horsepower.mean()

104.178391959799

In [8]:
df['horsepower']=df_horsepower.fillna(104)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    float64
 4   weight        398 non-null    float64
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
dtypes: float64(5), int64(3), object(1)
memory usage: 28.1+ KB


In [9]:

df['name'].unique()

array(['chevrolet chevelle malibu', 'buick skylark 320',
       'plymouth satellite', 'amc rebel sst', 'ford torino',
       'ford galaxie 500', 'chevrolet impala', 'plymouth fury iii',
       'pontiac catalina', 'amc ambassador dpl', 'dodge challenger se',
       "plymouth 'cuda 340", 'chevrolet monte carlo',
       'buick estate wagon (sw)', 'toyota corona mark ii',
       'plymouth duster', 'amc hornet', 'ford maverick', 'datsun pl510',
       'volkswagen 1131 deluxe sedan', 'peugeot 504', 'audi 100 ls',
       'saab 99e', 'bmw 2002', 'amc gremlin', 'ford f250', 'chevy c20',
       'dodge d200', 'hi 1200d', 'chevrolet vega 2300', 'toyota corona',
       'ford pinto', 'plymouth satellite custom', 'ford torino 500',
       'amc matador', 'pontiac catalina brougham', 'dodge monaco (sw)',
       'ford country squire (sw)', 'pontiac safari (sw)',
       'amc hornet sportabout (sw)', 'chevrolet vega (sw)',
       'pontiac firebird', 'ford mustang', 'mercury capri 2000',
       'opel 1900'

In [10]:
df.head()

mpg  cylinders  displacement  ...  model year  origin                       name
0  18.0          8         307.0  ...          70       1  chevrolet chevelle malibu
1  15.0          8         350.0  ...          70       1          buick skylark 320
2  18.0          8         318.0  ...          70       1         plymouth satellite
3  16.0          8         304.0  ...          70       1              amc rebel sst
4  17.0          8         302.0  ...          70       1                ford torino

[5 rows x 9 columns]

## 분류와 연속 컬럼 구분

In [11]:
df.head(8)

mpg  cylinders  displacement  ...  model year  origin                       name
0  18.0          8         307.0  ...          70       1  chevrolet chevelle malibu
1  15.0          8         350.0  ...          70       1          buick skylark 320
2  18.0          8         318.0  ...          70       1         plymouth satellite
3  16.0          8         304.0  ...          70       1              amc rebel sst
4  17.0          8         302.0  ...          70       1                ford torino
5  15.0          8         429.0  ...          70       1           ford galaxie 500
6  14.0          8         454.0  ...          70       1           chevrolet impala
7  14.0          8         440.0  ...          70       1          plymouth fury iii

[8 rows x 9 columns]

### 컬럼 형태 분류(check columns)
* 연속형: displacement, horsepower, weigth,acceleration, 
* 중립형: mpg,cylinders,origin
* 분류형: model year, name

(대게 카테고리형은 소숫점 및 값이 존재하지 않는다.)

standard scaler  : 연속형<br>
one hot incoding : 분류형 

하나의 데이터에 300개는 있어야 교육이된다
그래서 name은 교육시키기에 카운터 수가 작아서 빼고 진행

#### 중립형을 연속형인지 분륳형인지 판단

In [12]:
df['name'].value_counts()

ford pinto                      6
amc matador                     5
toyota corolla                  5
ford maverick                   5
amc hornet                      4
                               ..
vokswagen rabbit                1
dodge dart custom               1
dodge colt (sw)                 1
volkswagen 1131 deluxe sedan    1
cadillac eldorado               1
Name: name, Length: 305, dtype: int64

In [13]:
df['mpg'].describe(include='all')

count    398.000000
mean      23.514573
std        7.815984
min        9.000000
25%       17.500000
50%       23.000000
75%       29.000000
max       46.600000
Name: mpg, dtype: float64

예) count 398인데 mpg length가 129 값이 나왔다. 이때 분류형일까 연속형일까?

* 왼쪽카테고리에는 소숫점 이하 값 있어 이럴땐 연속형으로 분류할수 있어/  오른쪽 반복된카운트값

In [14]:
df['mpg'].value_counts()

13.0    20
14.0    19
18.0    17
15.0    16
26.0    14
        ..
34.2     1
38.1     1
37.2     1
32.1     1
32.7     1
Name: mpg, Length: 129, dtype: int64

In [15]:
df['cylinders'].describe()

count    398.000000
mean       5.454774
std        1.701004
min        3.000000
25%        4.000000
50%        4.000000
75%        8.000000
max        8.000000
Name: cylinders, dtype: float64

In [16]:
df['cylinders'].value_counts()

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

In [17]:
df['origin'].describe()

count    398.000000
mean       1.572864
std        0.802055
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max        3.000000
Name: origin, dtype: float64

In [18]:
df['origin'].value_counts()

1    249
3     79
2     70
Name: origin, dtype: int64

* 연속형: displacement, horsepower, weigth,acceleration, mpg

* 분류형: model year, name,cylinders,origin

label:mpg<br>
feature: others without name

## 시각화: 통계

### 정규화 단계:
 * standard scaler<br>
z = (x-u) / s <br>
z(모집단)<br>
x(샘플의 표준 점수)<br>
u(훈련 샘플의 평균)<br>
s(훈련 샘플의 표준 편차)<br>

    fit(X [, y, 샘플 _ 가중치])

In [19]:
Y = df['mpg']
X_continus = df[['displacement','horsepower', 'weight','acceleration']]
X_category = df[['model year','cylinders','origin']]

In [20]:
from sklearn import preprocessing

In [21]:
scaler = preprocessing.StandardScaler()
type(scaler)

sklearn.preprocessing._data.StandardScaler

In [22]:
#연속형만 적용해서 교육시킨다(패턴만 갖게 됨)
scaler.fit(X_continus)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [23]:
#패턴에 의해 값을 변환해서 X에 담는다
X = scaler.transform(X_continus)

In [24]:
from sklearn.linear_model import LinearRegression

In [25]:
#리니어 레그래이션 쌍둥이 만들기 lr
lr = LinearRegression()

type(lr)

sklearn.linear_model._base.LinearRegression

In [26]:
#해당 내용을 fitting한다, 교육끝
#fit은 하면 패턴만 갖게된다~ 1차방정식

lr.fit(X,Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [27]:
lr.score(X,Y)

0.7061364273740109

## predict로 전달 하기
X_continus = df[['displacement','horsepower', 'weight','acceleration']]

훈련시킬때 넣은 순서로 넣어줘야함

In [28]:
df.head(1)

mpg  cylinders  displacement  ...  model year  origin                       name
0  18.0          8         307.0  ...          70       1  chevrolet chevelle malibu

[1 rows x 9 columns]

In [29]:
#소비자에게 알려줘야 할 값이 결과로 나온다

#lr.predict([[307.0,130.0,3504.0,12.0]])

In [30]:
#scaler를 이미 해서 패턴을 만들어놨기 때문에 transform을 사용한다.
x_cusmter = scaler.transform([[307.0,130.0,3504.0,12.0]])
x_cusmter.shape

(1, 4)

In [31]:
#lr.predict에 넣어주기
lr.predict(x_cusmter)

array([18.99743588])

## 서비스 단계

pickle — Python object serialization

pickle은 바이너리 방식(wb) 파일로 저장이 된다. <br>
바이너리 사람이 익숙한 형태가 아니다.<br>
class만 담을수 있다.<br>


In [32]:
import pickle

In [33]:
pickle.dump(lr, open('./autompg_lr.pkl', 'wb'))

pickle이 저장 된 상태에서 저장된 autompg_lr.pkl을 다운로드한다

saves 폴더를 만들고 다운로드 했던 autompg_lr.pkl 을 끌어온다

pickle은 바이너리 방식(rb) 파일 불러온다.


In [36]:
!ls -l ./autompg_lr.pkl

-rw-r--r-- 1 root root 567 Jul  6 06:28 ./autompg_lr.pkl


In [38]:
pickle.load(open('./autompg_lr.pkl', 'rb'))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

* pickle로 상대 서비스개발자에게 scaler를 넘겨줄때 쓰는 방법

In [39]:
pickle.dump(scaler, open('./autompt_standardscaler.pkl', 'wb'))

서비스 개발자는 저장한 autompt_standardscaler.pkl를 다운로드해서 본인 일하는곳에 로드하기


## one hot encoding


In [41]:
X_category

model year  cylinders  origin
0            70          8       1
1            70          8       1
2            70          8       1
3            70          8       1
4            70          8       1
..          ...        ...     ...
393          82          4       1
394          82          4       2
395          82          4       1
396          82          4       1
397          82          4       1

[398 rows x 3 columns]

In [45]:
X_category['origin'].value_counts()

#카테고리
#1 , 2 , 3
#? | ? | ?
#1 | 0 | 0
#0 | 1 | 0
#0 | 0 | 1

1    249
3     79
2     70
Name: origin, dtype: int64

In [51]:
df_origin=pd.get_dummies(X_category['origin'], prefix='origin')

In [53]:
df_cylinders=pd.get_dummies(X_category['cylinders'],prefix='cylinders')

In [57]:
df_origin.shape, df_cylinders.shape

((398, 3), (398, 5))

In [59]:
# X_continus + df_cylinders + df_origin
X_continus.head(3)

displacement  horsepower  weight  acceleration
0         307.0       130.0  3504.0          12.0
1         350.0       165.0  3693.0          11.5
2         318.0       150.0  3436.0          11.0

컬럼명 잘 봐야 함, 서비스하려면 작업을해줘야함

In [65]:
#원한인코딩응로 붙히는 작업:pd.concat([X_continus, df_cylinders, df_origin], axis='columns')
X= pd.concat([X_continus, df_cylinders, df_origin], axis='columns')

In [66]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y)
X_train.shape,X_test.shape,Y_train.shape,Y_test.shape

((298, 12), (100, 12), (298,), (100,))

In [68]:
#좋은 스코어를 뽑아내기 위한 모델중의 하나가 xgboot(dicision tree 태생)
import xgboost

In [71]:
xgb= xgboost.XGBRegressor()
xgb

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [72]:
xgb.fit(X_train,Y_train)

[07:23:34] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [78]:
#교육시킨 데이터 train으로 스코어 확인
xgb.score(X_train,Y_train)

0.9303310644334573

In [79]:
#교육시키지 않은 데이터 test로 스코어 확인
xgb.score(X_test,Y_test)

0.7695531987593649